lib/index.js

```diff
@@ -543,23 +545,64 @@ class Config {
       return
     }
 
+    const cliWorkspaces = this[_get]('workspaces', 'cli')
+
     for (const p of walkUp(this.cwd)) {
-      // walk up until we have a nm dir or a pj file
-      const hasAny = (await Promise.all([
-        stat(resolve(p, 'node_modules'))
-          .then(st => st.isDirectory())
-          .catch(() => false),
-        stat(resolve(p, 'package.json'))
-          .then(st => st.isFile())
-          .catch(() => false),
-      ])).some(is => is)
-      if (hasAny) {
+      const hasNodeModules = await stat(resolve(p, 'node_modules'))
+        .then((st) => st.isDirectory())
+        .catch(() => false)
+
+      const hasPackageJson = await stat(resolve(p, 'package.json'))
+        .then((st) => st.isFile())
+        .catch(() => false)
+
+      if (!this.localPrefix && (hasNodeModules || hasPackageJson)) {
         this.localPrefix = p
-        return
+
+        // if workspaces are disabled, return now
+        if (cliWorkspaces === false) {
+          return
```

Comment:
Why is this inside the for loop?

Reply:
Oh it's so that we set localPrefix to the very first walkUp we find no matter what, fencepost issue which I'll back away from slowly cause I'm very bad at them.